cli/src/commands/publish.js

```diff
@@ -0,0 +1,216 @@
+const path = require('path')
+const { reporter, chalk } = require('@dhis2/cli-helpers-engine')
+const FormData = require('form-data')
+const fs = require('fs-extra')
+const inquirer = require('inquirer')
+const { createClient } = require('../lib/httpClient')
+const parseConfig = require('../lib/parseConfig')
+const makePaths = require('../lib/paths')
+
+const constructUploadUrl = appId => `/api/v1/apps/${appId}/versions`
+
+const isValidServerVersion = v => !!/(\d+)\.(\d+)/.exec(v)
+
+const dumpHttpError = (message, response) => {
+    if (!response) {
+        reporter.error(message)
+        return
+    }
+
+    reporter.error(
+        message,
+        response.status,
+        typeof response.data === 'object'
+            ? response.data.message
+            : response.statusText
+    )
+    reporter.debugErr('Error details', response.data)
+}
+
+const resolveBundle = (cwd, params) => {
+    const appBundle = {}
+    // use file-path from params
+    if (params.file) {
+        try {
+            const filePath = path.resolve(cwd, params.file)
+            if (!fs.statSync(filePath).isFile()) {
+                reporter.error(`${params.file} is not a file`)
+                process.exit(1)
+            }
+            appBundle.id = params.appId
+            appBundle.version = params.fileVersion
+            appBundle.path = filePath
+            appBundle.name = path.basename(filePath)
+        } catch (e) {
+            reporter.error(`File does not exist at ${params.file}`)
+            process.exit(1)
+        }
+    } else {
+        // resolve file from built-bundle
+        const paths = makePaths(cwd)
+        const builtAppConfig = parseConfig(paths)
+        if (!builtAppConfig.id) {
+            reporter.error(
+                `No App Hub id found for app. Add an ${chalk.bold(
+                    'id'
+                )}-field to ${chalk.bold('d2.config.js')}`
+            )
+            process.exit(1)
+        }
+        appBundle.id = builtAppConfig.id
+        appBundle.version = builtAppConfig.version
+        appBundle.path = path.relative(
+            cwd,
+            paths.buildAppBundle
+                .replace(/{{name}}/, builtAppConfig.name)
+                .replace(/{{version}}/, builtAppConfig.version)
+        )
+        appBundle.name = builtAppConfig.name
+
+        if (!fs.existsSync(appBundle.path)) {
+            reporter.error(
+                `App bundle does not exist, run ${chalk.bold(
+                    'd2-app-scripts build'
+                )} before deploying.`
+            )
+            process.exit(1)
+        }
+    }
+
+    return appBundle
+}
+
+const promptForConfig = async params => {
+    if (process.env.CI && (!params.apikey || !params.minVersion)) {
+        reporter.error(
+            'Prompt disabled in CI mode - missing apikey or minVersion parameter.'
+        )
+        process.exit(1)
+    }
+
+    const responses = await inquirer.prompt([
```

Comment:
(minor) should we also prompt for `id` in the case this isn't a platform app?

Reply:
Can add that, I do have `implies: 'appId` if `file`-param is present though, so you would need to provide the `appId` from param, but I guess prompting for that also is better.

Reply:
Minor, if you think it's OK as-is for now that's fine - there might be cases where the app is built with the platform but no appId has been specified, in which case it would be nice to prompt (and, even more nice, to offer to save the prompt input to the d2.config.js file - though that's not as easy since it's a dynamic file.